cranelift/codegen/src/ir/pcc.rs

```diff
@@ -373,9 +354,20 @@ impl FactContext {
     /// a memory access of the given size, is valid.
     pub fn check_address(&self, fact: &Fact, size: u32) -> PccResult<()> {
         match fact {
-            Fact::PointsTo {
-                region: MemoryRegion { max },
-            } => ensure!(u64::from(size) <= *max, OutOfBounds),
+            Fact::Mem { ty, offset } => {
+                let end_offset: i64 = offset
+                    .checked_add(i64::from(size))
+                    .ok_or(PccError::Overflow)?;
+                let end_offset: u64 =
+                    u64::try_from(end_offset).map_err(|_| PccError::OutOfBounds)?;
+                match &self.memory_types[*ty] {
+                    ir::MemoryTypeData::Struct { size, .. }
+                    | ir::MemoryTypeData::Memory { size } => {
+                        ensure!(end_offset <= *size, OutOfBounds)
+                    }
```

Comment:
This could be
```suggestion
                ensure!(
                    self.memory_types[*ty].size().map_or(false, |s| end_offset <= s),
                    OutOfBounds
                );
```

Reply:
I thought a bit about this, but I think on balance I like the current form better, as we'll need to go back to it once we have dynamic limits.